Refactor of Neutralizing Gas and multiple ability fixes

andrebastosdias opened the pull request:
### Edited: October 7th 2025

This ended up becoming a Neutralizing Gas refactor.

Now, no abilities use conditions except Flash Fire and the Gen 9 DLC1 abilities Protosynthesis and Quark Drive. By default, _**all**_ abilities run their `End` event except these with conditions (and Neutralizing Gas and a _commanding_ Pokémon). This also allows Flash Fire to be Skill Swapped by a Mold Breaker Pokémon and keep the condition.

---

Supersedes my very old PR [Automatically activate Booster Energy during weather/terrain changes](https://github.com/smogon/pokemon-showdown/pull/10832)

Protosynthesis/Quark Drive should react to Weather/Terrain changes, similarly to seeds.

---

Zen Mode was implemented as a condition before the `End` event was added to the ability: https://github.com/smogon/pokemon-showdown/commit/326dec34b8cff03b3d4634356ad739209f787c02. And yes, Zen Mode announces when Darmanitan reverts forms.

---

https://bulbapedia.bulbagarden.net/wiki/Flower_Gift_(Ability)
https://bulbapedia.bulbagarden.net/wiki/Forecast_(Ability)
This was described on both abilities' Bulbapedia pages, stating that they should revert form from Gen 5 onwards. I tested Gens 4, 5 and 8, and they reverted in both Gen 5 (with Gastro Acid) and Gen 8 (with Neutralizing Gas). Cherrim-Sunshine's stats are not boosted in Gen 4 if its ability is suppressed.

---

This all works well with the recent commit: [Reset ability and item state on switch in](https://github.com/smogon/pokemon-showdown/commit/5ead2874eb6c215978bd3e69d26de06ee2b4169a).

---

I removed abilities and items `End` events from running on switch out. Now, the abilities that need to run on switch out, implement the `SwitchOut` event, those being: extreme weathers, Air Lock/Cloud Nine, Ruin abilities and Neutralizing Gas.

---

Fixes: https://www.smogon.com/forums/threads/protosynthesis-should-never-deactivate-during-its-start-event.3774067/

andrebastosdias committed:
Cherrim and Castform should revert forms if their abilities are suppressed

urkerab commented on data/abilities.ts:
```diff
@@ -1443,6 +1450,12 @@ export const Abilities: import('../sim/dex-abilities').AbilityDataTable = {
 		},
+		onEnd(pokemon) {
+			if (pokemon.baseSpecies.baseSpecies !== 'Castform' || pokemon.transformed) return;
+			if (pokemon.baseSpecies.id !== 'castform') {
```
```suggestion
			if (pokemon.species.id !== 'castform') {
```
Also both abilities do check for `isActive` but in slightly different ways. I don't know why one checks for `hp` and the other does not though.

andrebastosdias replied:
It does, Castform hp check was at the end.

andrebastosdias committed:
Clean code

andrebastosdias committed:
Fix lint

andrebastosdias committed:
Don't revert if switching out

andrebastosdias committed:
Simplify Zen Mode

Also, Zen Mode shouldn't activate on switch out, so we add the pokemon.beingCalledBack check

andrebastosdias committed:
Add test

andrebastosdias committed:
Refactor Neutralizing Gas

andrebastosdias committed:
Revert Slow Start refactor

andrebastosdias committed:
Add Cud Chew to list of abilities

andrebastosdias commented:
This ended up becoming a Neutralizing Gas refactor. Some abilities, like Gorilla Tactics, were incorrectly not ending. Others also weren't ending, but that's mostly irrelevant since there are suppression checks elsewhere, such as with Air Lock and Cloud Nine.

From what I tested, all ability `End` events trigger when Neutralizing Gas enters the field, except for abilities that have specific conditions attached, such as Cud Chew, Flash Fire, Protosynthesis, Quark Drive and Unburden. In these cases, Neutralizing Gas simply suspends the conditions instead of removing them. The only exception is Slow Start (maybe shouldn't be implemented as a condition?).

andrebastosdias committed:
Update condition

andrebastosdias committed:
Fix Slow Start

andrebastosdias committed:
Merge branch 'master' into flowergift

andrebastosdias committed:
Merge remote-tracking branch 'upstream/master' into flowergift

andrebastosdias committed:
Refactor abilities

andrebastosdias committed:
Change condition order

andrebastosdias commented on data/abilities.ts:
```diff
@@ -1340,6 +1338,13 @@ export const Abilities: import('../sim/dex-abilities').AbilityDataTable = {
 		},
+		onEnd(pokemon) {
+			if (!pokemon.isActive || !pokemon.hp || pokemon.baseSpecies.baseSpecies !== 'Cherrim' ||
+				pokemon.transformed || pokemon.beingCalledBack) return;
```
I don't remember why I added `pokemon.beingCalledBack`

andrebastosdias replied:
The abilities should `End` with Neutralizing Gas, but shouldn't "end" when the Pokémon is just switching out.

andrebastosdias commented on data/mods/gen9dlc1/abilities.ts:
```diff
@@ -67,6 +75,21 @@ export const Abilities: import('../../../sim/dex-abilities').ModdedAbilityDataTa
+				pokemon.addVolatile('quarkdrive', pokemon, item);
+			}
+		},
+		// Implement as a condition so that it doesn't get suppressed by Neutralizing Gas
```
I thought it didn't have the flag. Isn't just the boost that isn't suppressed, but you still can't start the ability if Neutralizing Gas is on the field?

andrebastosdias replied:
> — v3.0.0 —
> Protosynthesis/Quark Drive are now affected by Neutralizing Gas ([DarkFE](https://www.smogon.com/forums/posts/9893355/))
> — v3.0.1 —
> Protosynthesis is now disabled while Cloud Nine is active ([amlodude](https://www.smogon.com/forums/posts/10374141/))

It seems that it might have been 2 different patches or just discovered at different times.

andrebastosdias committed:
Fix gen9dlc1 Paradox abilities

andrebastosdias committed:
Fix Protosynthesis start event

andrebastosdias committed:
Same fix for gen9dlc1

andrebastosdias committed:
Remove Supreme Overlord End event

andrebastosdias committed:
Change Tera Shift to a Start event

andrebastosdias committed:
Revert changes

andrebastosdias committed:
Merge branch 'master' into flowergift

andrebastosdias committed:
Merge branch 'master' into flowergift

andrebastosdias committed:
Don't send silent messages when switching out

andrebastosdias committed:
Merge branch 'master' into flowergift

andrebastosdias committed:
Merge remote-tracking branch 'upstream/master' into flowergift

andrebastosdias committed:
Merge remote-tracking branch 'upstream/master' into flowergift

andrebastosdias committed:
Use undefined for mod field removals

andrebastosdias committed:
Run 'SwitchOut' not 'End' events when switching out

andrebastosdias committed:
Merge remote-tracking branch 'upstream/master' into flowergift

andrebastosdias committed:
Comment Ruin abilities SwitchOut event

andrebastosdias committed:
Merge branch 'master' into flowergift
